codegen/templates.go

```diff
@@ -1,5 +1,258 @@
 package codegen
 
+import (
+	"fmt"
+	"io"
+	"io/ioutil"
+	"sort"
+	"strings"
+	"text/template"
+
+	"github.com/hashicorp/go-hclog"
+	"github.com/hashicorp/vault/sdk/framework"
+	"github.com/terraform-providers/terraform-provider-vault/util"
+)
+
+var (
+	// templateRegistry holds templates for each type of file.
+	templateRegistry = map[templateType]string{
+		// TODO in separate PR - add templateTypeDataSource
+		// TODO in separate PR - add templateTypeDoc
+		templateTypeResource: "/codegen/templates/resource.go.tpl",
+	}
+
+	// These are the types of fields that OpenAPI has that we support
+	// converting into Terraform fields.
+	supportedParamTypes = []string{
+		"array",
+		"boolean",
+		"integer",
+		"string",
+	}
+)
+
+func newTemplateHandler(logger hclog.Logger) (*templateHandler, error) {
+	// Read in the template for each template type in the registry and
+	// cache them to be used repeatedly.
+	templates := make(map[templateType]*template.Template, len(templateRegistry))
+	for tmplType, pathFromHomeDir := range templateRegistry {
+		pathToFile := pathToHomeDir + pathFromHomeDir
+		templateBytes, err := ioutil.ReadFile(pathToFile)
+		if err != nil {
+			return nil, err
+		}
+		t, err := template.New(tmplType.String()).Parse(string(templateBytes))
+		if err != nil {
+			return nil, err
+		}
+		templates[tmplType] = t
+	}
+	return &templateHandler{
+		logger:               logger,
+		templates:            templates,
+		templatableEndpoints: make(map[string]*templatableEndpoint),
+	}, nil
+}
+
+type templateHandler struct {
+	logger               hclog.Logger
+	templates            map[templateType]*template.Template
+	templatableEndpoints map[string]*templatableEndpoint
+}
+
+// Write takes one endpoint and uses a template to generate text
+// for it. This template is written to the given writer. It's exported
+// because it's the only method intended to be called by external callers.
+func (h *templateHandler) Write(wr io.Writer, tmplType templateType, parentDir string, endpoint string, endpointInfo *framework.OASPathItem) error {
+	templatable, ok := h.templatableEndpoints[endpoint]
+	if !ok {
+		// Since each endpoint will have a code file and a doc file, let's cache
+		// the template-friendly version of the endpoint so it doesn't have to be
+		// converted into that format twice.
+		t, err := h.toTemplatable(parentDir, endpoint, endpointInfo)
```

Comment:
Ah yes, good point. I didn't want to call it `tmpl` because then it sounds like a template, and that could be confusing because it's a custom struct. I ended up declaring the error just above so we can immediately use the (hopefully) best word, templatable.

https://github.com/terraform-providers/terraform-provider-vault/pull/746/commits/868dcdb2f0f9195066edf5b82d9677491bffbb00